src/common/cockpitwebinject.c

```diff
@@ -84,62 +87,68 @@ cockpit_web_inject_push (CockpitWebFilter *filter,
   const gchar *mark, *data, *pos;
   gsize mark_len, data_len, at;
   GBytes *bytes;
+  gsize written;
 
   /* Stop searching once injected */
-  if (self->inject)
+  mark = g_bytes_get_data (self->marker, &mark_len);
+  data = g_bytes_get_data (block, &data_len);
+  written = at = 0;
+
+  while (self->injected < self->maximum && at < data_len)
     {
-      mark = g_bytes_get_data (self->marker, &mark_len);
-      data = g_bytes_get_data (block, &data_len);
-      at = 0;
+      if (self->partial)
+        pos = data + at;
+      else
+        pos = memchr (data + at, mark[0], data_len - at);
 
-      while (at < data_len)
-        {
-          if (self->partial)
-            pos = data + at;
-          else
-            pos = memchr (data + at, mark[0], data_len - at);
+      /* Couldn't find the character anywhere? */
+      if (!pos)
+        break;
 
-          /* Couldn't find the character anywhere? */
-          if (!pos)
+      for (at = (pos - data); self->partial < mark_len && at < data_len; self->partial++, at++)
+        {
+          if (mark[self->partial] != data[at])
             break;
+        }
 
-          for (at = (pos - data); self->partial < mark_len && at < data_len; self->partial++, at++)
-            {
-              if (mark[self->partial] != data[at])
-                break;
-            }
+      /* Found a match */
+      if (self->partial == mark_len)
+        {
+          self->partial = 0;
 
-          /* Found a match */
-          if (self->partial == mark_len)
+          if (written < at)
             {
-              self->partial = 0;
-
-              bytes = g_bytes_new_from_bytes (block, 0, at);
+              bytes = g_bytes_new_from_bytes (block, written, at - written);
               function (func_data, bytes);
               g_bytes_unref (bytes);
+            }
 
-              function (func_data, self->inject);
-              g_bytes_unref (self->inject);
-              self->inject = NULL;
+          function (func_data, self->inject);
+          self->injected++;
 
-              bytes = g_bytes_new_from_bytes (block, at, data_len - at);
-              function (func_data, bytes);
-              g_bytes_unref (bytes);
-
-              block = NULL;
-              break;
-            }
+          written = at;
```

Comment:
couldn't this technically move into the `if (written < at)` block? https://github.com/cockpit-project/cockpit/pull/4113/files#diff-caaa68f1c7ba59d627b671504cabca99R123